devdoc/channel.md

```diff
@@ -0,0 +1,267 @@
+
+```mermaid
+stateDiagram-v2
+    B: EPOCH=x,STATE=CHANNEL_READY
+    N: EPOCH=x,STATE=CHANNEL_PUSHED
+    R: EPOCH=x,STATE=CHANNEL_PULLED
+    E: EPOCH=x+1,STATE=CHANNEL_READY
+
+    [*] --> B: channel_create
+
+    B --> N:  channel_push
+    B --> R:  channel_pull
+
+    N --> E:  channel_pull
+    R --> E: channel_push
+
+    N --> E:  cancel channel_push
+    R --> E: cancel channel_pull
+
+    R --> R: channel_pull
+    N --> N: channel_push
+
+    E --> [*]: channel_destroy
+```
+
+## Reentrancy
+
+`channel` is reentrant. This means that `channel_pull` and `channel_push` can be called from callbacks of this module.
+
+## Exposed API
+```c
+#define CHANNEL_RESULT_VALUES \
+    CHANNEL_RESULT_SYNC, \
+    CHANNEL_RESULT_ASYNC, \
+    CHANNEL_RESULT_REFUSED, \
+    CHANNEL_RESULT_INVALID_ARGS, \
+    CHANNEL_RESULT_ERROR
+
+MU_DEFINE_ENUM(CHANNEL_RESULT, CHANNEL_RESULT_VALUES);
+
+#define CHANNEL_CALLBACK_RESULT_VALUES \
+    CHANNEL_CALLBACK_RESULT_OK, \
+    CHANNEL_CALLBACK_RESULT_CANCELLED, \
+    CHANNEL_CALLBACK_RESULT_ABANDONED
+
+MU_DEFINE_ENUM(CHANNEL_CALLBACK_RESULT, CHANNEL_CALLBACK_RESULT_VALUES);
+
+#include "umock_c/umock_c_prod.h"
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+
+typedef void(*PULL_CALLBACK)(void* pull_context, THANDLE(RC_PTR) data, CHANNEL_CALLBACK_RESULT result);
+typedef void(*PUSH_CALLBACK)(void* push_context, CHANNEL_CALLBACK_RESULT result);
+typedef struct CHANNEL_TAG CHANNEL;
+
+THANDLE_TYPE_DECLARE(CHANNEL);
+
+    MOCKABLE_FUNCTION(, THANDLE(CHANNEL), channel_create)
+    MOCKABLE_FUNCTION(, CHANNEL_RESULT, channel_pull, THANDLE(CHANNEL), channel, PULL_CALLBACK, pull_callback, void*, pull_context, THANDLE(ASYNC_OP)*, out_op_pull);
+    MOCKABLE_FUNCTION(, CHANNEL_RESULT, channel_push, THANDLE(CHANNEL), channel, THANDLE(RC_PTR), data, PUSH_CALLBACK, push_callback, void*, push_context, THANDLE(ASYNC_OP)*, out_op_push);
+
+#ifdef __cplusplus
+}
+#endif /* __cplusplus */
+
+```
+
+### channel_create
+```c
+    MOCKABLE_FUNCTION(, CHANNEL_HANDLE, channel_create)
+```
+
+`channel_create` creates the channel and returns it.
+
+**SRS_CHANNEL_43_001: [** `channel_create` shall call `THANDLE_MALLOC` with `dispose` as `channel_dispose`. **]**
+
+**SRS_CHANNEL_43_056: [** `channel_create` shall set the `state` to `CHANNEL_READY` and `EPOCH` to `0` by calling `interlocked_exchange`. **]**
+
+**SRS_CHANNEL_43_002: [** If there are any failures, `channel_create` shall fail and return `NULL`. **]**
+
+### channel_dispose
+```c
+    void channel_dispose(CHANNEL* channel)
+```
+
+`channel_dispose` disposes the given `channel`.
+
+**SRS_CHANNEL_43_004: [** `channel_dispose` shall obtain the current state of the channel by calling `interlocked_add`. **]**
+
+**SRS_CHANNEL_43_057: [** If the current `state` is `CHANNEL_PULLING` or `CHANNEL_PUSHING`, `channel_dispose` shall wait for the state to change by calling `wait_on_address`. **]**
```

Comment:
>CHANNEL_PULLING [](http://example.com/codeflow?start=53&length=15)

I don't think this can ever happen, can it? 

For `channel_dispose` to be called (seeing as this is a THANDLE) would mean that there are no other references to it. 

And if there are no references... how come the state is somewhere "in between", as in "somebody **is calling**" pull or push.

Wondering if a unit test can even be written for this.

Reply:
Hmm yeah this won't happen